File: p1_python演習/ex15.py

# ex15.py
# 2018/02/19 Yuma Saito

import numpy as np

docs = [["リンゴ", "リンゴ"], ["リンゴ", "レモン"], ["レモン", "ミカン"]]
terms = ["リンゴ", "レモン", "ミカン"]

def tf(term, doc):
    """
    単語term、文書docに対するtf値を計算する。
    tf(term, doc) = (文書内に現れる単語termの数) / (文書内の総単語数)
    """
    count = 0   # 上式の分子の値を計算
    for word in doc:
        if word == term:
            count += 1
    return count / len(doc)

def idf(term, docs):
    """
    単語term、文書リストdocsに対するidf値を計算する。
    idf(term, docs) = log{(総文書数) / (単語termが含まれる文書数)} + 1
    """
    count = 0
    for doc in docs:
        if term in doc:
            count += 1
    return np.log10(len(docs) / count) + 1.

def tf_idf(terms, docs):
    """
    tf_idf[i, j]が文書i、単語jのtf-idf値を表しているような
    tf-idf行列を計算する
    """
    # [docs, terms]サイズの行列を生成
    tfidf = np.zeros([len(docs), len(terms)])
    for i, doc in enumerate(docs):
        for j, term in enumerate(terms):
            tfidf[i, j] = tf(term, doc) * idf(term, docs)
    return tfidf

print(tf_idf(terms, docs))

File: p1_python演習/cpp_answer/ex15.cpp
#include<iostream>
#include<string>
#include<vector>
#include<set>
#include<fstream>
#include<cmath>

double tf(const std::string& term, const std::vector<std::string>& doc) {
    int num = 0;
    for (const auto& w : doc) {
        if (w == term) {
            num++;
        }
    }
    return (double)num / doc.size();
}

double idf(const std::string& term, const std::vector<std::vector<std::string>>& docs) {
    int df = 0;
    for (const auto& doc : docs) {
        for (const auto& w : doc) {
            if (w == term) {
                df++;
                break;
            }
        }
    }
    return std::log10((double)docs.size() / df) + 1;
}

std::vector<std::vector<double>> tf_idf(const std::vector<std::string>& terms, const std::vector<std::vector<std::string>>& docs) {
    std::vector<std::vector<double>> result(docs.size(), std::vector<double>(terms.size()));
    for (int i = 0; i < docs.size(); i++) {
        for (int j = 0; j < terms.size(); j++) {
            result[i][j] = tf(terms[j], docs[i]) * idf(terms[j], docs);
        }
    }

    return result;
}

int main() {
    std::vector<std::vector<std::string>> docs {
        { "リンゴ", "リンゴ" },
        { "リンゴ", "レモン" },
        { "レモン", "ミカン" }
    };
    std::vector<std::string> terms {
        "リンゴ", "レモン", "ミカン"
    };

    auto tf_idf_mat = tf_idf(terms, docs);

    for (int i = 0; i < docs.size(); i++) {
        printf("%c", i ? ' ' : '[');
        printf("[");
        for (int j = 0; j < terms.size(); j++) {
            if (j) {
                printf("  ");
            }
            printf("%.10f", tf_idf_mat[i][j]);
        }
        if (i == docs.size() - 1) {
            printf("]");
        }
        printf("]\n");
    }
}
